DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/main.cpp

```diff
@@ -0,0 +1,430 @@
+  q.memcpy(in, in_pixels.data(), pixel_count * sizeof(PixelT)).wait();
+
+  // allocate space for the intensity sigma LUT
+  float* sig_i_lut_data_ptr = IntensitySigmaLUT::AllocateDevice(q);
+
+  // create the intensity sigma LUT data locally on the host
+  IntensitySigmaLUT sig_i_lut_host(params);
+
+  // copy the intensity sigma LUT to the device
+  sig_i_lut_host.CopyDataToDevice(q, sig_i_lut_data_ptr).wait();
+  //////////////////////////////////////////////////////////////////////////////
+
+  // track timing information in ms
+  std::vector<double> time(runs);
+
+  // print out some info
+  std::cout << "Runs:             " << runs << "\n";
+  std::cout << "Columns:          " << cols << "\n";
+  std::cout << "Rows:             " << rows << "\n";
+  std::cout << "Frames:           " << frames << "\n";
+  std::cout << "Filter Size:      " << kFilterSize << "\n";
+  std::cout << "Pixels Per Cycle: " << kPixelsPerCycle << "\n";
+  std::cout << "Maximum Columns:  " << kMaxCols << "\n";
+  std::cout << "\n";
+
+  try {
+    // run the design multiple times to increase the accuracy of the timing
+    for (int i = 0; i < runs; i++) {
+      // run ANR
+      time[i] =
+          RunANR(q, in, out, cols, rows, frames, params, sig_i_lut_data_ptr);
+
+      // Copy the output back from the device
+      q.memcpy(out_pixels.data(), out, pixel_count * sizeof(PixelT)).wait();
+
+      // validate the output on the last iteration
+      if (!kDisableDeviceMem && (i == (runs-1))) {
+        passed &= Validate(out_pixels.data(), ref_pixels.data(), rows, cols);
+      } else {
+        passed &= true;
+      }
+    }
+  } catch (exception const& e) {
+    std::cout << "Caught a synchronous SYCL exception: " << e.what() << "\n";
+    std::terminate();
+  }
+
+  // free the allocated device memory
+  sycl::free(in, q);
+  sycl::free(out, q);
+  sycl::free(sig_i_lut_data_ptr, q);
+
+  // write the output files if device memory was used (output is meaningless
+  // otherwise)
+  if constexpr (!kDisableDeviceMem) {
+    WriteOutputFile(data_dir, out_pixels, cols, rows);
+  }
+
+  // print the performance results
+  if (passed) {
+    // NOTE: when run in emulation, these results do not accurately represent
+    // the performance of the kernels in actual FPGA hardware
+    double avg_time_ms =
+        std::accumulate(time.begin() + 1, time.end(), 0.0) / (runs - 1);
+
+    size_t input_count_mega = pixel_count * frames * sizeof(PixelT) * 1e-6;
+
+    std::cout << "Execution time: " << avg_time_ms << " ms\n";
+    std::cout << "Throughput: " << (input_count_mega / (avg_time_ms * 1e-3))
+              << " MB/s\n";
+    std::cout << "PASSED\n";
+    return 0;
+  } else {
+    std::cout << "FAILED\n";
+    return 1;
+  }
+}
+
+// declare kernel and pipe names globally to reduce name mangling
+class ANRInPipeID;
+class ANROutPipeID;
+class InputKernelID;
+class OutputKernelID;
+
+//
+// Run the ANR algorithm on the FPGA
```

Comment:
Shouldn't this say 'on the target device', as it could be the emulator device (or in theory some non-FPGA target)?